service/controller/legacy/v22/resource/ipam/create.go

```diff
@@ -86,42 +85,44 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 
 			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("reused allocated cluster CIDR %#q", subnetCIDR))
 
-		} else {
-			r.logger.LogCtx(ctx, "level", "debug", "message", "allocating cluster subnet CIDR")
+			// Once we have all information together, regardless the update path, we
+			// update the CR status. Note that we try to use the latest resource version
+			// of the CR to get the status update properly sorted without any conflict.
+			{
+				r.logger.LogCtx(ctx, "level", "debug", "message", "updating CR status")
 
-			subnetCIDR, err = r.allocateSubnet(ctx)
-			if err != nil {
-				return microerror.Mask(err)
+				customResource.Status.Cluster.Network.CIDR = subnetCIDR.String()
+				customResource.Status.AWS.AvailabilityZones = statusAZs
+
+				_, err = r.g8sClient.ProviderV1alpha1().AWSConfigs(customResource.Namespace).UpdateStatus(&customResource)
+				if err != nil {
+					return microerror.Mask(err)
+				}
+
+				r.logger.LogCtx(ctx, "level", "debug", "message", "updated CR status")
+
+				r.logger.LogCtx(ctx, "level", "debug", "message", "canceling reconciliation")
+				reconciliationcanceledcontext.SetCanceled(ctx)
 			}
+		} else {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "allocating cluster subnet CIDR")
 
 			randomAZs, err := r.selectRandomAZs(key.SpecAvailabilityZones(customResource))
 			if err != nil {
 				return microerror.Mask(err)
 			}
 
-			statusAZs, err = splitSubnetToStatusAZs(subnetCIDR, randomAZs)
-			if err != nil {
-				return microerror.Mask(err)
+			callbacks := network.AllocationCallbacks{
+				GetReservedNetworks:     r.getReservedSubnets,
+				PersistAllocatedNetwork: r.persistAllocatedNetwork(customResource, randomAZs),
 			}
 
-			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("allocated cluster subnet CIDR %#q", subnetCIDR))
-		}
-
-		// Once we have all information together, regardless the update path, we
-		// update the CR status. Note that we try to use the latest resource version
-		// of the CR to get the status update properly sorted without any conflict.
-		{
-			r.logger.LogCtx(ctx, "level", "debug", "message", "updating CR status")
-
-			customResource.Status.Cluster.Network.CIDR = subnetCIDR.String()
-			customResource.Status.AWS.AvailabilityZones = statusAZs
-
-			_, err = r.g8sClient.ProviderV1alpha1().AWSConfigs(customResource.Namespace).UpdateStatus(&customResource)
+			subnetCIDR, err = r.networkAllocator.Allocate(ctx, r.networkRange, r.allocatedSubnetMask, callbacks)
 			if err != nil {
 				return microerror.Mask(err)
 			}
 
-			r.logger.LogCtx(ctx, "level", "debug", "message", "updated CR status")
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("allocated cluster subnet CIDR %#q", subnetCIDR))
```

Comment:
That isn't very clear why allocating subnet leads to reconciliation canceling. Maybe:

```suggestion
			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("updated CR status with allocated cluster subnet CIDR %#q", subnetCIDR))
```